pkg/server/plugin/plugin.go

```diff
@@ -116,55 +125,123 @@ type InitContext struct {
 	Config           skyconfig.Configuration
 }
 
-func (c *InitContext) AddPluginConfiguration(name string, path string, args []string) *Plugin {
+// AddPluginConfiguration creates and appends a plugin
+func (c *Context) AddPluginConfiguration(name string, path string, args []string) *Plugin {
 	plug := NewPlugin(name, path, args, c.Config)
 	c.plugins = append(c.plugins, &plug)
 	return &plug
 }
 
-func (c *InitContext) InitPlugins() {
-	for _, plug := range c.plugins {
-		plug.Init(c)
+// InitPlugins initializes all plugins registered
+func (c *Context) InitPlugins() {
+	wg := sync.WaitGroup{}
+	for _, eachPlugin := range c.plugins {
+		wg.Add(1)
+		go func(plug *Plugin) {
+			defer wg.Done()
+			plug.Init(c)
+		}(eachPlugin)
 	}
+
+	log.
+		WithField("count", len(c.plugins)).
+		Info("Wait for all plugin configurations")
+	wg.Wait()
+	c.SendEvents("before-plugins-ready", []byte{}, false)
+	c.SendEvents("after-plugins-ready", []byte{}, false)
 }
 
 // IsReady returns true if all the configured plugins are available
-func (c *InitContext) IsReady() bool {
-	for _, plug := range c.plugins {
-		if !plug.IsReady() {
+func (c *Context) IsReady() bool {
+	for _, eachPlugin := range c.plugins {
+		if !eachPlugin.IsReady() {
 			return false
 		}
 	}
 	return true
 }
 
-// Init instantiates a plugin. This sets up hooks and handlers.
-func (p *Plugin) Init(context *InitContext) {
-	p.transport.SetInitHandler(func(out []byte, err error) error {
-		if err != nil {
-			panic(fmt.Sprintf("Unable to get registration info from plugin. Error: %v", err))
+// SendEvents sends event to all plugins
+func (c *Context) SendEvents(name string, data []byte, async bool) {
```

Comment:
need to mention explicitly in the comment that `async` being `false` means the event messages are sent to each plugin one after or another, or that event messages are sent to all plugins at once, then wait for all operations to complete